internal/app/singularity/pull.go

```diff
@@ -177,71 +170,66 @@ func downloadImageCallback(totalSize int64, r io.Reader, w io.Writer) error {
 // OrasPull will download the image specified by the provided oci reference and store
 // it at the location specified by file, it will use credentials if supplied
 func OrasPull(name, ref string, force bool, ociAuth *ocitypes.DockerAuthConfig) error {
-	ref = strings.TrimPrefix(ref, "//")
-
-	spec, err := reference.Parse(ref)
-	if err != nil {
-		return fmt.Errorf("unable to parse oci reference: %s", err)
+	if !force {
+		if _, err := os.Stat(name); err == nil {
+			return fmt.Errorf("image file already exists: %q - will not overwrite", name)
+		}
 	}
 
-	// append default tag if no object exists
-	if spec.Object == "" {
-		spec.Object = SifDefaultTag
-		sylog.Infof("No tag or digest found, using default: %s", SifDefaultTag)
+	sysCtx := &ocitypes.SystemContext{
+		DockerAuthConfig: ociAuth,
 	}
 
-	credFn := func(_ string) (string, string, error) {
-		if ociAuth != nil {
-			return ociAuth.Username, ociAuth.Password, nil
-		}
-
-		return "", "", nil
+	sum, err := oras.ImageSHA(ref, sysCtx)
+	if err != nil {
+		return fmt.Errorf("failed to get checksum for %s: %s", ref, err)
 	}
-	resolver := docker.NewResolver(docker.ResolverOptions{Credentials: credFn})
 
-	wd, err := os.Getwd()
+	imageName := uri.GetName("oras:" + ref)
+
+	imagePath := cache.OrasImage(sum, imageName)
+	exists, err := cache.OrasImageExists(sum, imageName)
 	if err != nil {
-		return fmt.Errorf("failed to get working directory: %s", err)
+		return fmt.Errorf("unable to check if %s exists: %v", imagePath, err)
 	}
 
-	store := content.NewFileStore(wd)
-	defer store.Close()
+	if !exists {
+		sylog.Infof("Downloading image with ORAS")
 
-	store.AllowPathTraversalOnWrite = true
-	store.DisableOverwrite = !force
+		if err := oras.DownloadImage(imagePath, ref, ociAuth); err != nil {
```

Comment:
I suppose this is downloading the image to imagePath?

In that case, I would create a temporary filename _first_, download and verify, and only if that succeeds rename it to the final name in the cache. Also note that in case of error, the invalid file should be deleted.

1. assign temporary filename
2. download to temporary filename
3. verify temporary filename
4. if anything above fails, delete the temporary file, otherwise rename to the final name

If you move this logic to a function, you could do:

```
func downloadToCache(...) error {
    tmpfn := ...
    defer os.Remove(tmpfn)
    if err := download(src, tmpfn, ...) {
        return err // this deletes the temporary file if it was created
    }
    if err := verify(tmpfn, ...) {
        return err // this deletes the temporary
    }
    return os.Rename(tmpfn, finalfn) // if this fails, tmpfn gets deleted
}
```

Reply:
The current semantics for the cache is that if a corrupt image exists in the cache `<CacheType>ImageExists()` will indicate that it does not exist in the cache and it is to be overwritten. I think for this PR its best leave this the same between the different cache types, but since we are in the process of revisiting how we work with the cache in PRs like #3623 this should be the default workflow. I agree that it doesnt make sense to allow a corrupt image to be in the cache